[JAVA]: WIP - Java Tutorial: Getting Started Section (#12538)
apache/arrow: docs/source/java/install.rst

.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

Installing Java Modules
=======================

.. contents::

System Compatibility
--------------------

Java modules are regularly built and tested on macOS and Linux distributions.

Java Compatibility
------------------

Java modules are currently compatible with Java 8 / 9 / 10 / 11.

Installing from Maven
---------------------

By default, Maven will download from the central repository: https://repo.maven.apache.org/maven2/org/apache/arrow/

> Configure your pom.xml with java module version needed, for example for: memory-netty, format, and vector

Reviewer: This is something that would be good to explain (though probably elsewhere in the docs): what is the difference between memory-netty and memory-unsafe

Author: Yes, this is part of this ticket https://issues.apache.org/jira/browse/ARROW-15573, working on that

> Configure your pom.xml with java module version needed, for example for: memory-netty, format, and vector

Reviewer: ```suggestion
Configure your pom.xml with the Java modules needed, for example:
``arrow-memory-netty``, ``arrow-format``, and ``arrow-vector``.
```

        <modelVersion>4.0.0</modelVersion>
        <groupId>org.example</groupId>
        <artifactId>demo</artifactId>
        <version>1.0-SNAPSHOT</version>
        <properties>
            <arrow.version>7.0.0</arrow.version>
        </properties>
        <dependencies>
            <dependency>
                <groupId>org.apache.arrow</groupId>
                <artifactId>arrow-vector</artifactId>
                <version>${arrow.version}</version>
            </dependency>
            <dependency>
                <groupId>org.apache.arrow</groupId>
                <artifactId>arrow-memory-netty</artifactId>
                <version>${arrow.version}</version>
            </dependency>
            <dependency>
                <groupId>org.apache.arrow</groupId>
                <artifactId>arrow-format</artifactId>
                <version>${arrow.version}</version>
            </dependency>
        </dependencies>
    </project>

Installing from Source
----------------------

See :ref:`java-development`.

Installing Nightly Packages
---------------------------

.. warning::
    These packages are not official releases. Use them at your own risk.

> All arrow nightly version are uploaded to github assets for example for March 01 it is uploading to `Github Nightly`_

Reviewer: ```suggestion
Arrow nightly builds are uploaded to GitHub. For example, for 2022/03/01, they can be found at `Github Nightly`_.
```

> For example if you need to test your code with these artifacts, then, you need to configure maven settings with:

Reviewer: ```suggestion
To test your code with these artifacts, then configure Maven with:
```

.. code-block:: xml

    $ cat ~/.m2/settings.xml
    <?xml version="1.0" encoding="UTF-8"?>
    <settings xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.1.0 http://maven.apache.org/xsd/settings-1.1.0.xsd" xmlns="http://maven.apache.org/SETTINGS/1.1.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <profiles>
        <profile>
          <repositories>
            <repository>
               <id>staged</id>
               <name>staged-releases</name>
               <url>https://github.com/ursacomputing/crossbow/releases/tag/nightly-2022-03-01-0-github-java-jars/</url>
               <releases>
                 <enabled>true</enabled>
               </releases>
               <snapshots>
                 <enabled>true</enabled>
               </snapshots>
             </repository>
          </repositories>
          <id>arrownightly</id>
        </profile>
      </profiles>
    </settings>
    $ mvn -Parrownightly clean install -X
    Downloading from staged: https://github.com/ursacomputing/crossbow/releases/tag/nightly-2022-03-01-0-github-java-jars/org/apache/arrow/arrow-vector/8.0.0.dev143/arrow-vector-8.0.0.dev143.pom

> Consider that you could see arrow nightly builds at `Arrow Nightly`_.

Reviewer: ```suggestion
Arrow nightly builds are posted on the mailing list at `builds@arrow.apache.org`_.
```

> .. _Arrow Nightly: https://lists.apache.org/list.html?builds@arrow.apache.org

Reviewer: ```suggestion
.. _builds@arrow.apache.org: https://lists.apache.org/list.html?builds@arrow.apache.org
```

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

Installing Java Modules
=======================

System Compatibility
--------------------

Java modules are regularly built and tested on macOS and Linux distributions.

Java Compatibility
------------------

Java modules is currently compatible with JDK 8 / 9 / 10 / 11.

Using Maven
-----------

> Downloading in Maven is triggered by a project declaring a dependency that is not present in the local repository.

Author: Deleted

Central Repository
******************

By default, Maven will download from the central repository: https://repo.maven.apache.org/maven2/org/apache/arrow/

Configure your pom.xml with java module version needed. For example:

.. code-block::

    <dependency>
        <groupId>org.apache.arrow</groupId>
        <artifactId>arrow-vector</artifactId>
        <version>7.0.0</version>

>     </dependency

Reviewer: There's a missing bracket here, also, shouldn't we show the 'full' set of dependencies needed? i.e. memory-netty, format, and vector? We also don't need to explain Maven or pom.xml here. Just state what modules are needed and show the example XML.

Author: Added libraries and xml demo

Configure your maven settings.xml to download artifacts from staging repository with:

.. code-block:: xml

    $ cat ~/.m2/settings.xml
    <?xml version="1.0" encoding="UTF-8"?>
    <settings xsi:schemaLocation="http://maven.apache.org/SETTINGS/1.1.0 http://maven.apache.org/xsd/settings-1.1.0.xsd" xmlns="http://maven.apache.org/SETTINGS/1.1.0"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <profiles>
        <profile>
          <repositories>
            <repository>
               <id>staged</id>
               <name>staged-releases</name>
               <url>https://repository.apache.org/content/repositories/staging/</url>
               <releases>
                 <enabled>true</enabled>
               </releases>
               <snapshots>
                 <enabled>true</enabled>
               </snapshots>
             </repository>
          </repositories>
          <id>arrowrc</id>
        </profile>
      </profiles>
    </settings>
    $ mvn -Parrowrc clean install -X
    Downloading from staged: https://repository.apache.org/content/repositories/staging/

Installing from source
----------------------

See :ref:`java-development`.

> Installing Nightly Packages

Reviewer: If these are the nightly packages, then what is the purpose of the staging repository above?

Author: Nightly packages are added by arrow nightly jobs and uploaded to github assets.

For staging it is not aligned with github assets and it depend of dev/release/06-java-upload.sh.

If RC on nexus is closed, the artifacts will be available in the staging repository.

Reviewer: I still don't see why a developer should care about staging artifacts. Either they want the latest stable release or they want to test nightlies.

Author: Clear, deleted staging maven section

---------------------------

.. warning::
    These packages are not official releases. Use them at your own risk.

All arrow nightly version are uploaded to github assets for example for March 01 it is uploading to `Github Nightly`_

> .. image:: img/java_nightly_github_asset.png

Author: Deleted
